File: benchmark/baselines/typescript-strict/registration_service.ts
import type { Clock, RegistrationStore } from "./capabilities.js";
import type { Event, Person, Registration, RegistrationError, Result } from "./domain.js";
export function register(event: Event, person: Person, store: RegistrationStore, clock: Clock): Result<Registration, RegistrationError> {
  if (store.find(event.id, person.id) !== undefined) return { ok: false, error: "already_registered" };
  if (event.state === "closed") return { ok: false, error: "registration_closed" };
  if (event.registered >= event.capacity) return { ok: false, error: "event_full" };
  const registration = { eventId: event.id, personId: person.id, createdAt: clock.now() };
  return store.save(registration).ok ? { ok: true, value: registration } : { ok: false, error: "store_failed" };
}

File: benchmark/baselines/rust/registration_service.rs
use crate::capabilities::{Clock, RegistrationStore};
use crate::domain::{Event, EventState, Person, Registration, RegistrationError};
pub fn register<S: RegistrationStore, C: Clock>(event: &Event, person: &Person, store: &mut S, clock: &mut C) -> Result<Registration, RegistrationError> {
    if store.find(event.id, person.id).is_some() { return Err(RegistrationError::AlreadyRegistered); }
    if matches!(event.state, EventState::Closed) { return Err(RegistrationError::RegistrationClosed); }
    if event.registered >= event.capacity { return Err(RegistrationError::EventFull); }
    let registration = Registration { event_id: event.id, person_id: person.id, created_at: clock.now() };
    store.save(registration.clone()).map_err(|_| RegistrationError::StoreFailed)?;
    Ok(registration)
}
